[Quant] [Inductor] Enable QLinear weight prepack when input dimension size exceeds 2

leslie-fang-intel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113928
* #113912
* #113733

**Summary**
Enable the qlinear weight prepack when input dimension size exceeds 2. There are extra reshape node before and after the `addmm` or `mm` node if input dimension size exceeds 2.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k input_dim_exceeds_2
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

leslie-fang-intel committed:
[Quant] [Inductor] Enable QLinear weight prepack when input dimension size exceeds 2

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant] [Inductor] Enable QLinear weight prepack when input dimension size exceeds 2"


**Summary**
Enable the qlinear weight prepack when input dimension size exceeds 2. There are extra reshape node before and after the `addmm` or `mm` node.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k input_dim_exceeds_2
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant] [Inductor] Enable QLinear weight prepack when input dimension size exceeds 2"


**Summary**
Enable the qlinear weight prepack when input dimension size exceeds 2. There are extra reshape node before and after the `addmm` or `mm` node.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k input_dim_exceeds_2
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant] [Inductor] Enable QLinear weight prepack when input dimension size exceeds 2"


**Summary**
Enable the qlinear weight prepack when input dimension size exceeds 2. There are extra reshape node before and after the `addmm` or `mm` node if input dimension size exceeds 2.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k input_dim_exceeds_2
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant] [Inductor] Enable QLinear weight prepack when input dimension size exceeds 2"


**Summary**
Enable the qlinear weight prepack when input dimension size exceeds 2. There are extra reshape node before and after the `addmm` or `mm` node if input dimension size exceeds 2.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k input_dim_exceeds_2
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant] [Inductor] Enable QLinear weight prepack when input dimension size exceeds 2"


**Summary**
Enable the qlinear weight prepack when input dimension size exceeds 2. There are extra reshape node before and after the `addmm` or `mm` node if input dimension size exceeds 2.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k input_dim_exceeds_2
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant] [Inductor] Enable QLinear weight prepack when input dimension size exceeds 2"


**Summary**
Enable the qlinear weight prepack when input dimension size exceeds 2. There are extra reshape node before and after the `addmm` or `mm` node if input dimension size exceeds 2.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k input_dim_exceeds_2
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel commented on torch/_inductor/ir.py:
```diff
@@ -4837,37 +4837,23 @@ def _prepare_linear_fusion_create(
-        x.get_dtype(),
-        convert_shape_to_inductor(output_size),
-        convert_shape_to_inductor(output_stride),
+    kernel_layout = FlexibleLayout(
```
- Thanks for the comment. Change it back to use `FixedLayout`.
- Previously, we meet a issue of `convert_shape_to_inductor` with dynamic shape input (after changing the implementation in this PR) https://github.com/pytorch/pytorch/blob/a911b4db9d82238a1d423e2b4c0a3d700217f0c1/torch/_inductor/utils.py#L236-L238 and I found `LinearUnary` uses the `FlexibleLayout` which doesn't have this issue. That's why I also try to use `FixedLayout`. https://github.com/pytorch/pytorch/blob/81f93991d3be078fada6dcaf939c2fb0bdb67e63/torch/_inductor/ir.py#L5237-L5245

jgong5 replied:
Also, I found you cleaned up the fake_tensor trace. Should we clean up similar code in other related ops?

leslie-fang-intel replied:
Sure, Add a TODO comment in `_prepare_convolution_fusion_create`. Since `_prepare_convolution_fusion_create` is shared usage by several operators as `ConvolutionUnary`, `ConvolutionBinary`, `ConvolutionBinaryInplace`, `QConvPointWisePT2E` and so on. I prefer to do it in the following up PRs.

leslie-fang-intel committed:
Update on "[Quant] [Inductor] Enable QLinear weight prepack when input dimension size exceeds 2"


**Summary**
Enable the qlinear weight prepack when input dimension size exceeds 2. There are extra reshape node before and after the `addmm` or `mm` node if input dimension size exceeds 2.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k input_dim_exceeds_2
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel commented:
Hi @jerryzh168 @eellison, could you kindly help to take a look of this PR?

leslie-fang-intel committed:
Update on "[Quant] [Inductor] Enable QLinear weight prepack when input dimension size exceeds 2"


**Summary**
Enable the qlinear weight prepack when input dimension size exceeds 2. There are extra reshape node before and after the `addmm` or `mm` node if input dimension size exceeds 2.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k input_dim_exceeds_2
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel commented:
@pytorchbot merge
